src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -329,33 +354,143 @@ def frobenius_charpoly(self, var='X', algorithm='crystalline'):
+                return self._frobenius_charpoly.change_variable_name(var)
             self._frobenius_charpoly = getattr(self, method_name)(var)
             return self._frobenius_charpoly
         raise NotImplementedError(f'algorithm \"{algorithm}\" not implemented')
 
-    def _frobenius_charpoly_crystalline(self, var):
+    @cached_method
+    def _frobenius_charpoly_CSA(self, var='X'):
+        r"""
+        Return the characteristic polynomial of the Frobenius
+        endomorphism using Crystalline cohomology.
+
+        The algorithm works for Drinfeld `\mathbb{F}_q[T]`-modules of
+        any rank.
+
+        This method is private and should not be directly called.
+        Instead, use :meth:`frobenius_charpoly` with the option
+        `algorithm='CSA'`.
+
+        INPUT:
+
+        - ``var`` (default: ``'X'``) -- the name of the second variable
+
+        EXAMPLES::
+
+            sage: Fq = GF(5)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(3)
+            sage: phi = DrinfeldModule(A, [z^i for i in range(1, 50)])
+            sage: phi.frobenius_charpoly(algorithm="CSA")  # indirect doctest
+            X^48 + 4*X^47 + 4*X^46 + X^45 + 3*X^44 + X^42 + 4*X^41 + 4*X^39 + 2*X^37 + 4*X^36 + 3*X^35 + 2*X^33 + (4*T + 2)*X^32 + (4*T + 1)*X^31 + (3*T + 1)*X^30 + 4*T*X^29 + 4*T*X^28 + 2*X^27 + X^26 + (3*T + 4)*X^25 + (4*T + 4)*X^24 + (T + 1)*X^23 + (T + 3)*X^22 + 4*T*X^21 + (2*T + 1)*X^20 + 4*X^19 + 4*T*X^18 + (T + 4)*X^17 + 4*T^2*X^16 + (T^2 + 3*T + 3)*X^15 + (4*T^2 + 3*T + 3)*X^14 + (3*T^2 + 3*T + 3)*X^13 + (3*T^2 + 4*T + 4)*X^12 + (T^2 + 2*T + 2)*X^11 + (3*T^2 + 4*T + 4)*X^10 + (3*T^2 + T + 1)*X^9 + (4*T + 4)*X^8 + (3*T + 3)*X^7 + (T^2 + 3*T + 3)*X^6 + (3*T^2 + T + 1)*X^5 + (2*T^2 + 3*T + 3)*X^4 + (2*T^2 + 3*T + 3)*X^3 + 3*T^2*X^2 + 4*T^2*X + 2*T^3 + T + 1
+
+        ALGORITHM:
+
+        Compute the characteristic polynomial of the Frobenius from
+        the reduced characteristic polynomial of the Ore polynomial
+        `phi_T`. This algorithm is particularly interesting when the
+        rank of the Drinfeld module is large compared to the degree
+        of the extension `K/\mathbb F_q`.
+        """
+        E = self._base
+        EZ = PolynomialRing(E, name='Z')
+        n = self._base_degree_over_constants
+        phi_T = self.gen()
+        t = self.ore_variable()
+        rows = [ ]
+        # Compute the reduced charpoly
+        for i in range(n):
+            m = phi_T.degree() + 1
+            row = [EZ([phi_T[jj] for jj in range(j, m, n)]) for j in range(n)]
+            rows.append(row)
+            phi_T = t * phi_T
+        chi = Matrix(rows).charpoly()
+        # We format the result
+        K = self.base_over_constants_field()
+        A = self.function_ring()
+        r = self.rank()
+        lc = chi[0][r]
+        coeffs = [ A([K(chi[i][j]/lc).in_base()
+                      for i in range((r-j)*n // r + 1)])
+                   for j in range(r+1) ]
```

Comment:
```suggestion
        coeffs = [A([K(chi[i][j]/lc).in_base()
                     for i in range((r-j)*n // r + 1)])
                  for j in range(r+1)]
```
Did I just correct your pep8?!

Reply:
Yes 😭
(But it was Xavier's code...)